crates/partition-store/src/partition_store.rs

```diff
@@ -962,3 +988,120 @@ pub(crate) trait StorageAccess {
+            Self::KEY_KIND.serialize(bytes);
+        }
+
+        fn serialize_to<B: BufMut>(&self, bytes: &mut B) {
+            Self::serialize_key_kind(bytes);
+            bytes.put_u32(self.len() as u32);
+            bytes.put_slice(self.as_bytes());
+        }
+
+        fn deserialize_from<B: Buf>(bytes: &mut B) -> crate::partition_store::Result<Self> {
+            let key_kind = KeyKind::deserialize(bytes)?;
+            assert_eq!(key_kind, Self::KEY_KIND);
+
+            let len = bytes.get_u32() as usize;
+            let mut string_bytes = Vec::with_capacity(len);
+            bytes.copy_to_slice(&mut string_bytes);
+            Ok(String::from_utf8(string_bytes).expect("valid key"))
+        }
+
+        fn serialized_length(&self) -> usize {
+            KeyKind::SERIALIZED_LENGTH + self.len()
+        }
+    }
+
+    #[restate_core::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn concurrent_writes_and_reads() -> googletest::Result<()> {
+        let rocksdb = RocksDbManager::init(Constant::new(CommonOptions::default()));
+        let partition_store_manager =
+            PartitionStoreManager::create(Constant::new(StorageOptions::default()), &[]).await?;
+        let mut partition_store = partition_store_manager
+            .open_partition_store(
+                PartitionId::MIN,
+                PartitionKey::MIN..=PartitionKey::MAX,
+                OpenMode::CreateIfMissing,
+                &RocksDbOptions::default(),
+            )
+            .await?;
+        let number_writers: u32 = 10_000;
+        let key_a = "a".to_owned();
+        let key_b = "b".to_owned();
+
+        // put the initial values
+        partition_store.put_kv_raw(key_a.clone(), 0_u32.to_be_bytes())?;
+        partition_store.put_kv_raw(key_b.clone(), 0_u32.to_be_bytes())?;
+
+        let mut writer_tasks: FuturesUnordered<_> = (1..=number_writers)
+            .map(|writer_task| {
+                TaskCenter::spawn_unmanaged(TaskKind::TestRunner, "concurrent-writes", {
+                    let mut partition_store = partition_store.clone();
+                    let key_a = key_a.clone();
+                    let key_b = key_b.clone();
+
+                    async move {
+                        let mut txn = partition_store
+                            .transaction_with_isolation(IsolationLevel::RepeatableReads);
+                        let value_a = txn.get_kv_raw(key_a.clone(), decode_u32)?;
+                        let value_b = txn.get_kv_raw(key_b.clone(), decode_u32)?;
+
+                        assert_eq!(value_a, value_b, "both values must be equal");
+
+                        txn.put_kv_raw(key_a.clone(), writer_task.to_be_bytes())?;
+                        txn.put_kv_raw(key_b.clone(), writer_task.to_be_bytes())?;
+                        txn.commit().await?;
+
+                        Ok::<_, anyhow::Error>(())
+                    }
+                })
+                .expect("not shut down")
+            })
```

Comment:
I suppose this test can be written in a more deterministic way and do less work at the same time (faster). If we test that
1- put key1, key2 = valueA
2-create transaction with repeatable read isolation
3- read key1 from tx, should be valueA
4- outside the tx, set key2=valueB and commit the write batch
5- from tx, read key2. it should be valueA.
6- Profit. No loops.

Reply:
Very valid point. This is currently too complicated. Will change it.